### ISSUE TITLE
nn.functional.linear() for sparse tensor

### ISSUE THREAD

**Issue by weiyangfb** [pytorch/pytorch#12308]
Add `nn.functional.linear()` with autograd support for sparse tensor when input is sparse, while weights and bias are dense.

- autograd support for `addmm(D, S, D) -> D`
- add `sparse_matmul(S, D) -> D` with autograd support, also need to support broadcasting..

**Comment by lihuan19891123** [comment]
How about nn.functional.linear() with autograd for sparse tensor when input is dense, while weights is sparse?  In my situation, calculating grad for sparse weights is needed, but seems this is not supported by pytorch yet. I am sorry if I was wrong.

**Comment by weiyangfb** [comment]
@lihuan19891123 sure, that's doable too! Can I get a pointer to your model on github? Curious to know what applications out there need sparse :)

**Comment by fkhawar** [comment]
> @lihuan19891123 sure, that's doable too! Can I get a pointer to your model on github? Curious to know what applications out there need sparse :)
@weiyangfb I would also like the support for sparse weights as @lihuan19891123 mentioned . This would be useful in defining a sparse structure in the network connections e.g. a neuron would only connect to a subset of neurons from the previous layer.

**Comment by spfohl** [comment]
Should this now work with the changes to the Sparse API made in v1.0.0? From some prototyping, that appears to be the case, but I still run in to issues if a transpose operation is used.

**Comment by AlliedToasters** [comment]
@weiyangfb  If you are looking for a use case for linear with dense input vector, sparse weights matrix, please look at our [Sparse Evolutionary Training project](https://github.com/alliedtoasters/synapses).
We are currently using third-party libraries to fill these feature gaps but official support would be awesome. Plus, it can be faster I'm sure!

**Comment by souryadey** [comment]
I would really like Pytorch to support dense inputs and sparse weights. This is what our work on [pre-defined sparsity](https://ieeexplore.ieee.org/document/8689061) is based on. Thus far, our code uses sparse masking to zero out the sparse weights after every update. But the updates are still dense, so this is not the most efficient solution.

**Comment by yf225** [comment]
We need to discuss how we should prioritize this feature.

**Comment by souryadey** [comment]
I did some analysis on torch.sparse.mm vs torch.mm. Turns out that the sparse case is only faster when nnz is very low, like <= 1.5% of total elements. My findings are here:
https://medium.com/@souryadey/sparse-matrices-in-pytorch-be8ecaccae6

Seems like the COO format is not the best way to go.

**Comment by ezyang** [comment]
Unfortunately we don't have anyone in the core team who will be working on this in the near future. We'd be happy to accept a patch for adding some of these sparse matrix multiply combinations.

Changing the format from COO would be a bit more involved, we would have to hash out a design first.
